### DISCUSSION TITLE
eXO SSO configuration for Keycloak

### DISCUSSION THREAD

**Discussion by queza85** [exoplatform/exo-community#46] [Q&A]
Hi, I have a question that is setting for keycload based on SAML2
When we setup the keylocak configuration, we have issue to setup the password.

I put in correct keystorepass but it is shown error 'incorrect password' like capture.

First capture is shown that the password is working fine.
Second capture is error log even I used same password.

![image](https://github.com/exoplatform/exo-community/assets/2861131/82d24ec1-26ea-4d11-8d96-4a6a9f2360d9)

![image](https://github.com/exoplatform/exo-community/assets/2861131/67383ff4-9625-4f12-807f-c7055923d2df)

Can I get any advise to solve this?

**Comment by rdenarie** [comment]
Hello @queza85 
Can you provide configuration properties you use (not only password one but all saml2 related ? 
You can obfuscate private informations.

Thanks

**Reply by queza85** [reply]
Hi @rdenarie 
I use exo.properties as follow config

# SSO
gatein.sso.enabled=true
gatein.sso.saml.sp.enabled=true
gatein.sso.callback.enabled=true
gatein.sso.login.module.enabled=true
gatein.sso.login.module.class=org.gatein.sso.agent.login.SAML2IntegrationLoginModule
gatein.sso.valve.enabled=true
gatein.sso.valve.class=org.gatein.sso.saml.plugin.valve.ServiceProviderAuthenticator
gatein.sso.filter.login.sso.url=/portal/dologin
gatein.sso.filter.initiatelogin.enabled=false
gatein.sso.filter.logout.enabled=true
gatein.sso.filter.logout.class=org.gatein.sso.saml.plugin.filter.SAML2LogoutFilter
gatein.sso.filter.logout.url=${gatein.sso.sp.url}?GLO=true
gatein.sso.saml.config.file=/opt/exo/conf/saml2/keycloak-sp.xml

# Custom properties
gatein.sso.sp.host=http://10.1.20.78:8080
gatein.sso.sp.url=${gatein.sso.sp.host}/portal/dologin
gatein.sso.idp.host=https://kc.pos.com.my
gatein.sso.idp.url=${gatein.sso.idp.host}/realms/FELLOW/protocol/saml

gatein.sso.idp.alias=fellowtest
gatein.sso.idp.signingkeypass=peler123
gatein.sso.idp.keystorepass=peler123
gatein.sso.picketlink.keystore=/opt/exo/conf/saml2/exo-keystoretest.jks

**Comment by rdenarie** [comment]
Can you provide all the stack trace ? It seems that the very last part is missing.
Thanks

**Reply by queza85** [reply]
sure. I added log file when I restart the server with saml configuration. 
another file is keylocak saml file.

[platform-20231022-0.log](https://github.com/exoplatform/exo-community/files/13073598/platform-20231022-0.log)
[keycloak-sp.xml.txt](https://github.com/exoplatform/exo-community/files/13073603/keycloak-sp.xml.txt)

**Comment by jancrot** [comment]
I'm in the same team with queza. We're using Docker version

- Initially we thought it's wrong path to keystore file, but when we changed the path to wrong path we got another error saying file not found. This indicates that the path field in the configuration is indeed used
- We tried to remove the KeyStorePass from picketlink-sp.xml and we got another error saying the password is null which indicates the KeyStorePass field is indeed used
- We tried to write the values directly in picketlink-sp.xml, still same "Keystore password was incorrect" error shows up
- We tried to use different character for password in picketlink-sp.xml like single quote, double quote etc. Same error
- We tried to mask the password as mentioned here https://developer.jboss.org/docs/DOC-15382 still same error
- We read picketlink source code in github and it looks like there's no special thing in the code. It's basically just reading the URL and password and then pass it to Java native library to read
- We suspected this is a problem with JDK here https://bugs.openjdk.org/browse/JDK-8278989 so we wrote a simple Java program to read the keystore using the keystorepass. Compile the program inside container and run it there, with correct password there is no error. When we supplied wrong password to the simple program, we indeed got the same error
- We tried to create other keystore with different groupname like secp256r1, still same error. We use keytool and the command is something like this "keytool -genkey -groupname secp256r1 -keyalg EC -alias fellowtest -keystore exo-keystoretest.jks". We tried with keyalg RSA but can't find a way to modify the groupname
- We tried to create a cert using OpenSSL and then import it to keystore, still same error. We use keytool and the command is something like this "keytool -v -importkeystore -srckeystore keystore.p12 -srcstoretype PKCS12 -destkeystore exo-keystoretest.jks -destalias fellowtest -srcalias fellowtest"

We're stuck hehe. Any help would be appreciated

**Comment by rdenarie** [comment]
Hi,
Thanks for all information.
Can you provide the command you use to create the keystore file ?
Thanks

**Reply by queza85** [reply]
Hi, use like this. Cany ou confirm it? @jancrot  
> keytool -importkeystore -srckeystore exo-keystore.jks -srcstoretype PKCS12 -srcstorepass peler123 -destkeystore exo-keystore2.jks -deststoretype PKCS12 -deststorepass peler123

**Reply by jancrot** [reply]
We tried many, such as:
- keytool -genkey -alias fellowtest -keyalg RSA -keystore exo-keystoretest.jks
- keytool -genkey -groupname secp256r1 -keyalg EC -alias fellowtest -keystore exo-keystoretest.jks
- (importing from Open SSL) keytool -v -importkeystore -srckeystore keystore.p12 -srcstoretype PKCS12 -destkeystore exo-keystoretest.jks -destalias fellowtest -srcalias fellowtest

**Reply by rdenarie** [reply]
Ok.
The idea with the keystore is to store the IDP certificate into, so that, eXo can read and validate signature of assertions coming from the IDP.

In an old exo documentation about saml (we need to migrate in the new one), we had this 
![image](https://github.com/exoplatform/exo-community/assets/807839/d9a71f6d-62e0-4aa8-a868-ed2f68045903)

This step explain how to import the 'idp-certificate.crt' into the jbid_test_keystore.jks which exists in conf/saml2/
As I can see with your commands, you only create a keystore, and generating a key but you not add the IDP certificate into.

So you need to adapt your command to
- add your idp certificate into the existing jks keystore
- create your own keystore and add your idp certificate into

**Comment by queza85** [comment]
Hi 
@jancrot was found that keystore type is different.
Our own type is PKCS12. So he changed as JKS type an then it works fine.
Thanks all to help this on.
